### PR TITLE
User DMS PR

### PR DESCRIPTION
**Pull request by ankitpws**
# Description
These recommendations are intended to promote code quality and team communication during software development. They cover a variety of topics, including ensuring that pull requests are submitted to the correct branch, documenting new methods, preserving consistency across many services, and avoiding typical blunders like accessing APIs or DB queries within loops. Sensitive data should not be uploaded, and changes to environment variables or database models should be executed consistently. Teams may work more effectively and develop higher-quality software by adhering to these standards.

## Type of change
Please choose appropriate options.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Enhancement (additive changes to improve performance)
- [ ]  This change requires a documentation update

## Checklist 

- [ ]  It's critical to avoid making needless file modifications in contributions, such as adding new lines, console logs, or additional spaces, to guarantee cleaner and more efficient code. Furthermore, eliminating unnecessary imports from a file might enhance code readability and efficiency.
- [ ]  Ensure that the pull request is assigned to the right base branch and that the development branch name contains the JIRA Task Id. Furthermore, each commit message should include the JIRA Task Id in the manner "ED-100: message".
- [ ]  Only update packages if it is mentioned and authorized in the design document, and make sure that you have the required permissions.
- [ ]  Avoid making API and database queries inside a loop as it can lead to performance issues and slow down the system.
- [ ]  When calling another function inside a given function, add comments explaining the purpose and meaning of the passed arguments and expected return values.
- [ ]  If adding a blank argument in a function, add a comment explaining the reason for the blank argument.
- [ ]  Before submitting a pull request, do a self-review of your code to ensure there are no conflicts with the base branch and all comments have been addressed.
- [ ]  Before merging a pull request, it's important to have other team members review it to catch any potential errors or issues
- [ ]  To maintain code integrity, it's important to remove all related changes when removing code during a code review.
- [ ]  If new constants, endpoints, or utility functions are introduced, it is important to check if they already exist in the service to avoid any duplication.
- [ ]  Whenever a new environment variable is added to a service, it's important to ensure that the necessary changes are made to related files such as ".env.sample" and "envVariables.js" to maintain consistency and avoid errors. Additionally, the new environment variable should be added to the devops repository to ensure that it is properly documented and accessible to the team.
- [ ]  When adding a new function to a service, it is important to document it with relevant information such as the name, parameters, and return value in a consistent format across all services. Additionally, if there are any changes to the API response, ensure that the documentation in the controllers is updated accordingly. 
- [ ]  Write a clear and concise commit message that describes the changes made.
- [ ]  Maintain consistent function signature and code across all services when adding a function to multiple services. Implement changes to database models in all services that use the same model. 
- [ ]  Use only let and const. Do not use var.
- [ ]  Make common functions for repetitive code blocks.
- [ ]  Avoid uploading sensitive information such as secret tokens or passwords in pull requests to ensure data security.
- [ ]  Maintain consistent indentation and spacing throughout the code.

### CODE REVIEW

File: module/userProjects/helper.js [hunk 1]
```diff
@@ -8,507 +8,658 @@
...
+        "iteration": 1
+      }
+    }
+   * @returns {Promise} success Data.
+   */
+  static userDelete(userDeleteEvent) {
+    return new Promise(async (resolve, reject) => {
+      let userId = userDeleteEvent.edata.userId;
+      let userProfileUpdateData = [];
+      let userProjectData = await projectQueries.projectDocument(
```
Comment (aks30) [inline module/userProjects/helper.js:106]: @ankitpws This can all be in one query

File: module/userProjects/helper.js [hunk 2]
```diff
@@ -8,507 +8,658 @@
...
+        rawEvent.edata.type = CONSTANTS.common.USER_DELETE_TYPE;
+        rawEvent.edata.props = [];
+        let userObject = {
+          id: userId,
+          type: "User",
+        };
+        rawEvent.actor = userObject;
+        rawEvent.object = userObject;
+
+        let telemetryEvent = {
```
Comment (aks30) [inline module/userProjects/helper.js:166]: @ankitpws Make this an another utils function

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 18 Nov

File: .env.sample [hunk 3]
```diff
@@ -40,3 +40,8 @@ CERTIFICATE_SERVICE_URL = http://registry-service:8081
 
 PROJECT_CERTIFICATE_ON_OFF = "ON/OFF"                                           // Project certificate enable or disable flag
 
+USER_DELETE_TOPIC = {{env_name}}.delete.user                                    // Topic name for user delete event consumer
+
+ID = {{env_name}}.sunbird.ml.projects.service                                     // ID of service
```
[the remark this answers is not in the document]
Comment (ankitpws) [reply .env.sample]: changed to instance name

File: databaseQueries/projects.js [hunk 4]
```diff
@@ -91,40 +85,59 @@ module.exports = class Projects {
...
+  }
 
+  /**
+   * bulkUpdate function is used to do bulk operations.
+   * @method
+   * @name bulkUpdate
+   * @param {Object} [projectData] - project Data along with filter and quries.
+   * @returns {Object} - status of bulk operation.
+   */
+  static bulkUpdate(Data) {
```
[the remark this answers is not in the document]
Comment (ankitpws) [reply databaseQueries/projects.js]: using updateMany function

File: envVariables.js [hunk 5]
```diff
@@ -8,159 +8,205 @@
...
+  },
+  USER_DELETE_TOPIC: {
+    message: "Required user delete kafka topic to consume user delete event",
+    optional: false,
+  },
+  ID: {
+    message: "Required Service Id for communication with other services",
+    optional: false,
+  },
+  TELEMETRY_TOPIC: {
```
[the remark this answers is not in the document]
Comment (ankitpws) [reply envVariables.js]: added

File: generics/helpers/utils.js [hunk 6]
```diff
@@ -332,49 +334,90 @@ function noOfElementsInArray(data, filter = {}) {
...
+  });
+}
+
+/**
+ * generate telemetry raw event
+ * @function
+ * @name getTelemetryEvent
+ * @returns {Object} returns uuid.
+ */
+function getTelemetryEvent() {
```
[the remark this answers is not in the document]
Comment (ankitpws) [reply generics/helpers/utils.js]: done

File: generics/helpers/utils.js [hunk 7]
```diff
@@ -332,49 +334,90 @@ function noOfElementsInArray(data, filter = {}) {
...
+  let telemetryEvent = {
+    eid: "",
+    ets: epochTime(),
+    ver: "3.0",
```
[the remark this answers is not in the document]
Comment (ankitpws) [reply generics/helpers/utils.js]: done

File: generics/helpers/utils.js [hunk 8]
```diff
@@ -332,49 +334,90 @@ function noOfElementsInArray(data, filter = {}) {
...
+    actor: {},
+    context: {
+      channel: "",
+      pdata: {
+        id: process.env.ID,
+        pid: "manage-learn",
+        ver: packageData.version,
+      },
+      env: "",
+      cdata: [{ id: generateUUId(), type: "Request" }],
```
[the remark this answers is not in the document]
Comment (ankitpws) [reply generics/helpers/utils.js]: removed

File: generics/helpers/utils.js [hunk 9]
```diff
@@ -332,49 +334,90 @@ function noOfElementsInArray(data, filter = {}) {
...
+    eid: "",
+    ets: epochTime(),
+    ver: "3.0",
+    mid: generateUUId(),
+    actor: {},
+    context: {
+      channel: "",
+      pdata: {
+        id: process.env.ID,
+        pid: "manage-learn",
```
[the remark this answers is not in the document]
Comment (ankitpws) [reply generics/helpers/utils.js]: not passig from here deriving in the fly

File: generics/kafka/consumers/userDMS.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+ *   value:'{"eid":"BE_JOB_REQUEST","ets":1619527882745,"mid":"LP.1619527882745.32dc378a-430f-49f6-83b5-bd73b767ad36","actor":{"id":"delete-user","type":"System"},"context":{"channel":"01309282781705830427","pdata":{"id":"org.sunbird.platform","ver":"1.0"},"env":"dev"},"object":{"id":"<deleted-userId>","type":"User"},"edata":{"organisationId":"0126796199493140480","userId":"a102c136-c6da-4c6c-b6b7-0f0681e1aab9","suggested_users":[{"role":"ORG_ADMIN","users":["<orgAdminUserId>"]},{"role":"CONTENT_CREATOR","users":["<contentCreatorUserId>"]},{"role":"COURSE_MENTOR","users":["<courseMentorUserId>"]}],"action":"delete-user","iteration":1}}'
+ * }
+ * @returns {Promise} return a Promise.
+ */
+
+var messageReceived = function (message) {
+  return new Promise(async function (resolve, reject) {
+    try {
+      let parsedMessage = JSON.parse(message.value);
+      if (parsedMessage.edata.action === "delete-user") {
```
[the remark this answers is not in the document]
Comment (ankitpws) [reply generics/kafka/consumers/userDMS.js]: done

File: generics/kafka/consumers/userDMS.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
+/**
+ * name : userDMS.js
```
[the remark this answers is not in the document]
Comment (ankitpws) [reply generics/kafka/consumers/userDMS.js]: done

File: generics/kafka/producers.js [hunk 12]
```diff
@@ -3,83 +3,122 @@
...
+    : "OFF";
+const projectSubmissionTopic =
+  process.env.PROJECT_SUBMISSION_TOPIC &&
+  process.env.PROJECT_SUBMISSION_TOPIC != "OFF"
+    ? process.env.PROJECT_SUBMISSION_TOPIC
+    : "sl-improvement-project-submission-dev";
+const telemetryEventTopic = process.env.TELEMETRY_TOPIC
+  ? process.env.TELEMETRY_TOPIC
+  : "sl-telemetry-dev";
+const userDeleteTopic = process.env.USER_DELETE_TOPIC;
```
[the remark this answers is not in the document]
Comment (ankitpws) [reply generics/kafka/producers.js]: done

File: module/userProjects/helper.js [hunk 13]
```diff
@@ -8,507 +8,658 @@
...
+   * @returns {Promise} success Data.
+   */
+  static userDelete(userDeleteEvent) {
+    return new Promise(async (resolve, reject) => {
+      let userId = userDeleteEvent.edata.userId;
+      let userProfileUpdateData = [];
+      let userProjectData = await projectQueries.projectDocument(
+        {
+          userId: userId,
+          userProfile: { $exists: true },
```
[the remark this answers is not in the document]
Comment (ankitpws) [reply module/userProjects/helper.js]: not passing filter

File: module/userProjects/helper.js (hunk 1 above)
Comment (ankitpws) [reply module/userProjects/helper.js]: done with updateMany

File: module/userProjects/helper.js (hunk 2 above)
Comment (ankitpws) [reply module/userProjects/helper.js]: created new function

File: .env.sample [hunk 14]
```diff
@@ -40,3 +40,11 @@ CERTIFICATE_SERVICE_URL = http://registry-service:8081
 
 PROJECT_CERTIFICATE_ON_OFF = "ON/OFF"                                           // Project certificate enable or disable flag
 
+USER_DELETE_TOPIC = {{env_name}}.delete.user                                    // Topic name for user delete event consumer
```
Comment (aks30) [inline .env.sample:43]: @ankitpws - Let's make this dependent and configurable ON/OFF based

File: envVariables.js [hunk 15]
```diff
@@ -5,162 +5,179 @@
...
+     message: "Required Service ID",
+     optional: false,
+   },
+   "TELEMETRY_ON_OFF":{
+     message: "Required telemetry on/off status",
+     optional: false,
+   },
+   "TELEMETRY_TOPIC": {
+     message: "Required telemetry topic",
+     optional: false,
```
Comment (aks30) [inline envVariables.js:78]: @ankitpws This is not mandatory if TELEMETRY is OFF

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 22 Nov

File: config/kafka.js [hunk 16]
```diff
@@ -51,6 +52,9 @@ const connect = function() {
       process.env.KAFKA_URL
     );
 
+    // user Delete Consumer
+    _sendToKafkaConsumers(USER_DELETE_TOPIC, process.env.KAFKA_URL, USER_DELETE_ON_OFF);
```
Comment (aks30) [inline config/kafka.js:56]: @ankitpws We can if condition here

File: envVariables.js [hunk 17]
```diff
@@ -5,162 +5,185 @@
...
+   },
+   "TELEMETRY_ON_OFF":{
+     message: "Required telemetry on/off status",
+     optional: false,
+     default: "ON"
+   },
+   "TELEMETRY_TOPIC": {
+     message: "Required telemetry topic",
+     optional: true,
```
Comment (aks30) [inline envVariables.js:84]: @ankitpws Make this like Kafka

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 24 Nov

File: .env.sample (hunk 14 above)
Comment (ankitpws) [reply .env.sample:45]: made

File: config/kafka.js (hunk 16 above)
Comment (ankitpws) [reply config/kafka.js]: done

File: envVariables.js (hunk 17 above)
Comment (ankitpws) [reply envVariables.js:89]: done

**Pull request merged**